notebooks/triplet_loss_cifar10.py

```diff
@@ -0,0 +1,236 @@
+"""
+This notebooks borrows from https://www.tensorflow.org/addons/tutorials/losses_triplet and is intended to compare tf.addons triplet loss
+implementation against this one. It is also aimed at benchmarking the impact of the distance function.
+"""
+from pathlib import Path
+from pprint import pprint
+
+import itertools
+import io
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow.keras.callbacks import TensorBoard
+from tensorflow.keras.layers import Conv2D, Dense, Dropout, GlobalMaxPooling2D, Input, Flatten, MaxPooling2D, Lambda
+from tensorflow.keras.models import Sequential
+
+from keras_fsl.layers import GramMatrix
+from keras_fsl.losses.gram_matrix_losses import binary_crossentropy, class_consistency_loss, triplet_loss
+from keras_fsl.metrics.gram_matrix_metrics import classification_accuracy
+from keras_fsl.utils.tensors import get_dummies
+
+
+#%% Build datasets
+train_dataset, val_dataset, test_dataset = tfds.load(
+    name="cifar10", split=["train[:90%]", "train[90%:]", "test"], as_supervised=True
+)
+train_dataset = train_dataset.shuffle(1024).batch(64, drop_remainder=True)
+val_dataset = val_dataset.shuffle(1024).batch(64, drop_remainder=True)
+test_dataset = test_dataset.batch(64, drop_remainder=True)
+
+train_labels = [batch[1].numpy().tolist() for batch in train_dataset]
+val_labels = [batch[1].numpy().tolist() for batch in val_dataset]
+test_labels = [batch[1].numpy().tolist() for batch in test_dataset]
+train_steps = len(train_labels)
+val_steps = len(val_labels)
+test_steps = len(test_labels)
+
+print(
+    pd.concat(
+        [
+            pd.DataFrame({"label": tf.nest.flatten(train_labels)}).assign(split="train"),
+            pd.DataFrame({"label": tf.nest.flatten(val_labels)}).assign(split="val"),
+            pd.DataFrame({"label": tf.nest.flatten(test_labels)}).assign(split="test"),
+        ]
+    )
+    .groupby("split")
+    .apply(lambda group: pd.get_dummies(group.label).agg("sum"))
+)
+
+output_dir = Path("logs") / "triplet_loss_cifar10"
+output_dir.mkdir(exist_ok=True, parents=True)
+results = []
+
+#%% Save test labels for later visualization in projector https://projector.tensorflow.org/
+out_m = io.open(output_dir / "meta.tsv", "w", encoding="utf-8")
+for img, labels in tfds.as_numpy(test_dataset):
+    [out_m.write(str(x) + "\n") for x in labels]
```

Comment:
on en a déjà parlé non ? une boucle for nesté serait mieux. Ou alors il y a peut être un write_liines ou un truc du genre.
Ou encore : `'\n'.join(map(str, labels))`.
Une autre option pourrait être d'utiliser pandas.

Reply:
c'est copié/collé du notebook tensorflow je ne l'avais même pas lu en détail